### PR TITLE
Allow multiple TMs for dbkvs 

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Proof of concept for creation of multiple transaction managers with oracle. Still need to deal with 

- closing of the pool in a reasonable way -- potentially just ignoring close if reuse is true? 
- concurrent get ranges executor, probably similarly to the existing kvs executor case

### CODE REVIEW

**Comment by gmaretic** [comment]
Also, we probably want to have some sort of fairness guarantees for the connection pool similar to what we have with the executor views

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/config/ConnectionConfig.java [hunk 1]
```diff
@@ -74,9 +74,15 @@ public int getMinConnections() {
 
     @Value.Default
     public int getMaxConnections() {
+        // todo (gmaretic): this needs to be bumped if we are reusing the connection pool
```
Comment (gmaretic) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/config/ConnectionConfig.java:77]: This is not an actual todo, just a reminder that this config needs to be changed when we reuse -- will remove later

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java [hunk 2]
```diff
@@ -55,6 +55,7 @@
 
     private final ConnectionConfig connConfig;
     private final HikariConfig hikariConfig;
+    private final boolean closeable;
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java:58]: this is fine, though I think we should probably have some kind of wrapper indicating ownership (see `Ownable<T>`)? fine for testing, maybe not ideal for prod

if bandwidth allows we should get some pairing time tomorrow

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagers.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * limitations under the License.
+ */
+
+package com.palantir.nexus.db.pool;
+
+import com.palantir.nexus.db.pool.config.ConnectionConfig;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public final class HikariClientPoolConnectionManagers {
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagers.java:23]: should we use a factory object instead? In general statics are not great

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java [hunk 4]
```diff
@@ -39,9 +40,13 @@ public String tablePrefix() {
 
     @Value.Default
     public int poolSize() {
+        // todo (gmaretic): this should not be 64 if we are creating O(100) kvs instances as we could easily starve
+        // the shared cached threadpool
```
Comment (jeremyk-91) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java:44]: 👍

**Review by jeremyk-91 — COMMENTED** [review]
Looks about right in terms of mechanical changes! We might want to refine some bits for prod, but this is generally what I expected. There may be some things we should factor a bit differently.

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagers.java (hunk 3 above)
Comment (gmaretic) [reply atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagers.java:23]: Alternatives ended up messy

**Comment by gmaretic** [comment]
Ready for review, pending tests

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/ConnectionWithCallback.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package com.palantir.nexus.db.pool;
+
+import com.google.common.reflect.Reflection;
+import java.sql.Connection;
+
+public final class ConnectionWithCallback {
+    private ConnectionWithCallback() {
+        // sdfgj
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/ConnectionWithCallback.java:24]: nit: wat 🦆

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/ConnectionWithCallback.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.nexus.db.pool;
+
+import com.google.common.reflect.Reflection;
+import java.sql.Connection;
+
+public final class ConnectionWithCallback {
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/ConnectionWithCallback.java:22]: probably leave a note that the callback runs after the connection was closed.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/spi/SharedResourcesConfig.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface SharedResourcesConfig {
+    int sharedGetRangesPoolSize();
+
+    int sharedKvsExecutorSize();
+
+    LocalConnectionConfig connectionConfig();
```
Comment (jeremyk-91) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/spi/SharedResourcesConfig.java:27]: Is this common to all KVSes? Feels a bit like a DbKVS specific thing: I'm fine with us not doing the specific sub-impls right now, but we should probably leave a note here that this is used for relational key-value-services

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    private static final int GET_RANGES_QUEUE_SIZE_WARNING_THRESHOLD = 1000;
+
+    private GetRangesExecutors() {
+        // boo
+    }
+
+    @SuppressWarnings("DangerousThreadPoolExecutorUsage")
+    public static ExecutorService createGetRangesExecutor(
+            int numThreads, String prefix, Optional<Integer> sharedExecutorThreads) {
+        String name = prefix + "-get-ranges";
```
Comment (jeremyk-91) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java:42]: note to reader: this is not a function of namespace, so this is actually going to be shared safely.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        return wrap(SharedFixedExecutors.createOrGetShared(name, numThreads, sharedExecutorThreads));
+    }
+
+    private static ExecutorService wrap(ExecutorService executor) {
```
Comment (jeremyk-91) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java:46]: eh, can we say what this does? `wrapWithSizeMonitoring` or something?

File: atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagerView.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Override
+    public void close() throws SQLException {
+        // ignore
+    }
+
+    @Override
+    public void closeUnchecked() {
+        // ignore
+    }
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagerView.java:87]: Makes sense, but might want to make clear somewhere in docs of this class (maybe at the constructor) that we do *not* take ownership of the underlying object.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+                    queueSizeEstimate.getAndDecrement();
+                    command.run();
+                });
+            }
+
+            private void sanityCheckAndIncrementQueueSize() {
+                int currentSize = queueSizeEstimate.getAndIncrement();
+                if (currentSize >= GET_RANGES_QUEUE_SIZE_WARNING_THRESHOLD) {
+                    warningLogger.log(logger -> logger.warn(
+                            "You have {} pending getRanges tasks. Please sanity check both your level "
```
Comment (jeremyk-91) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java:89]: nit: if we're changing this, just make this say 'many'?

**Review by jeremyk-91 — APPROVED** [review]
👍  This seems to satisfy our requirements.

As noted in Goals, we aren't currently ever closing the hikari connection pool, though that's fine as long as we are very clear in javadocs that we don't take ownership of the object.

There are also some bits around the shared KVS config that are a bit DbKVS-specific, I find, but in the interest of time I'm OK with this.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/spi/SharedResourcesConfig.java (hunk 7 above)
Comment (gmaretic) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/spi/SharedResourcesConfig.java:27]: It actually can be. As it's currently wired, Cassandra KVS will actually wire the get ranges executors to be shared if this is set, but there is no reason we can't in the future wire up other ones if that is useful

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java (hunk 9 above)
Comment (gmaretic) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/GetRangesExecutors.java:46]: sure

File: atlasdb-dbkvs-hikari/src/test/java/com/palantir/nexus/db/pool/ConnectionWithCallbackTest.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    @Test
+    public void callbackRunsOnlyOnceWhenConnectionIsClosed() {
+        Connection connectionWithCallback = ConnectionWithCallback.wrap(connection, counter::incrementAndGet);
+
+        assertThat(counter).hasValue(0);
+
+        ExecutorService executor = PTExecutors.newFixedThreadPool(10);
+        CountDownLatch latch = new CountDownLatch(1);
+        List<Future<Void>> results = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/test/java/com/palantir/nexus/db/pool/ConnectionWithCallbackTest.java:50]: this is fine though in general I normally use 64 or 256 for this kind of test 😅

File: atlasdb-dbkvs-hikari/src/test/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagerViewTest.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Test
+    public void closingReturnedConnectionAllowsNewConnectionsToBeBorrowed() throws SQLException {
+        when(sharedManager.getConnection()).thenReturn(mock(Connection.class));
+        ConnectionManager connectionManager = new HikariClientPoolConnectionManagerView(sharedManager, 1, 0);
+
+        try (Connection connection = connectionManager.getConnectionUnchecked()) {
+            assertGetConnectionTimesOut(connectionManager);
+        } catch (SQLException throwables) {
+            fail("This was unexpected");
```
Comment (jeremyk-91) [inline atlasdb-dbkvs-hikari/src/test/java/com/palantir/nexus/db/pool/HikariClientPoolConnectionManagerViewTest.java:66]: sneaky! but fine for test :)

**Review by jeremyk-91 — APPROVED** [review]
👍

**Pull request merged**